File: packages/watchmen-analysis-client/src/model/suggestedAction.ts

export type ActionExecutionMode = 'auto' | 'manual' | 'approval';
export type ActionRiskLevel = 'low' | 'medium' | 'high' | 'critical';
export type ActionPriority = 'low' | 'medium' | 'high';

export interface ActionTypeParameter {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'email' | 'url';
  required: boolean;
  description?: string;
}

export interface ActionType {
  id: string;
  name: string;
  code: string; // e.g. 'notification', 'email', 'policy_adjust'
  description?: string;
  requiresApproval?: boolean;
  enabled: boolean;
  category: string; // e.g. 'Notification', 'Policy Operation'
  parameters?: ActionTypeParameter[];
}

export interface SuggestedActionCondition {
  metricName: string;
  operator: string; // '>', '<', '>=', '<=', '==', '!='
  value: string | number;
}

export interface SuggestedAction {
  id: string;
  name: string;
  typeId: string; // Reference to ActionType
  riskLevel: ActionRiskLevel;
  description: string;
  expectedOutcome?: string;
  conditions: SuggestedActionCondition[]; 
  executionMode: ActionExecutionMode;
  priority: ActionPriority;
  enabled: boolean;
  
  // Stats
  executionCount?: number;
  successRate?: number;
  lastExecuted?: string;
  
  parameters?: Record<string, any>;
}

File: packages/watchmen-model/src/watchmen_model/alert/suggested_action.py
from enum import Enum
from typing import List, Optional, Any, Dict, Union

from watchmen_model.common import SuggestedActionId, ActionTypeId, TenantId, UserBasedTuple
from watchmen_utilities import ExtendedBaseModel


class ActionExecutionMode(str, Enum):
	AUTO = 'auto'
	MANUAL = 'manual'
	APPROVAL = 'approval'


class ActionRiskLevel(str, Enum):
	LOW = 'low'
	MEDIUM = 'medium'
	HIGH = 'high'
	CRITICAL = 'critical'


class ActionPriority(str, Enum):
	LOW = 'low'
	MEDIUM = 'medium'
	HIGH = 'high'


class SuggestedActionCondition(ExtendedBaseModel):
	metricName: str = None
	operator: str = None
	value: Union[str, int, float] = None


class SuggestedAction(UserBasedTuple):
	suggestedActionId: SuggestedActionId = None
	name: str = None
	typeId: ActionTypeId = None
	riskLevel: ActionRiskLevel = None
	description: str = None
	expectedOutcome: Optional[str] = None
	conditions: List[SuggestedActionCondition] = None
	executionMode: ActionExecutionMode = None
	priority: ActionPriority = None
	enabled: bool = True

	# Stats
	executionCount: int = 0
	successRate: float = 0.0
	lastExecuted: Optional[str] = None

	parameters: Optional[Dict[str, Any]] = None
